03/arr-to-bst.md

```diff
@@ -0,0 +1,221 @@
+
+Please note one is not required to implement a self-balancing Binary Search Tree in order to solve this exercise. 
+
+It is recommended to break the problem down recursively by first setting the root of the Binary Search Tree to the middle element of the array.
+
+<br>
+<details style="max-width: 700px; margin: auto;">
+<summary>Click here to see the tests that will be run against your code</summary>
+
+```py
+def test_will_return_balanced_bst_for_odd_lengthed_list(self):
+    # Arrange
+    arr = [5, 10, 15, 20, 25, 30, 35, 40, 45]
+
+    # Act
+    result = arr_to_bst(arr)
+
+    # Assert
+    self.assertEqual(result.val, 25)
+    self.assertTrue(self.is_bst(result))
+    self.assertTrue(self.is_balanced_tree(result))
+
+def test_will_return_balanced_bst_for_even_lengthed_list(self):
+    # Arrange
+    arr = [1, 3, 9, 27, 81, 243]
+
+    # Act
+    result = arr_to_bst(arr)
+
+    # Assert
+    self.assertTrue(self.is_bst(result))
+    self.assertTrue(self.is_balanced_tree(result))
+
+def test_will_return_balanced_bst_for_long_list(self):
+    # Arrange
+    arr = []
+    num = 0
+    while num < 100:
+        arr.append(num)
+        num += 1
+
+    # Act
+    result = arr_to_bst(arr)
+
+    # Assert
+    self.assertTrue(self.is_bst(result))
+    self.assertTrue(self.is_balanced_tree(result))
+
+def test_will_return_none_for_empty_list(self):
+    # Arrange
+    arr = []
+
+    # Act
+    result = arr_to_bst(arr)
+
+    # Assert
+    self.assertEqual(result, None)
+```
+
+</details>
+
+## Prompts
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: paragraph
+* id: d0a2663c-8528-41b6-8e21-333812b18876
+* title: Ask Clarifying Questions
+* topics: pse
+##### !question
+
+List three or more questions whose answers would clarify the problem statement.
+
+For each question, provide an explanation which includes the effect your decision would have on how you would approach the problem.
+
+##### !end-question
+
+##### !explanation
+
+Here are some example clarifying questions:
+
+1. What kind of data is stored in the Tree Nodes?
+2. Is the array guaranteed to be non-empty?
+
+##### !end-explanation
+
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 2 -->
+<!-- prettier-ignore-start -->
+
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: 275e4d6e-accf-4dd3-89c4-9f23729615e8
+* title: Write Unit Tests
+* topics: pse
+##### !question
+
+1. Use the comments provided to write at least two example input/outputs:
+    * Consider at least one nominal and one edge case.
+    * What is the expected output for the given input?
+    * You can use the examples provided in the prompt, or other examples.
+2. Write unit tests for `arr_to_bst` for the nominal and edge cases you identified in the first step.
+
+*Note: Click the **Run Tests** button to save your tests for instructor feedback. No real tests are actually run again your unit tests.*
```

Comment:
Hmm... should we provide them with `is_bst`, `height` and `is_balanced_tree` here? We may need to rethink how we frame the question for this IPQ since the tests are a bit more involved.

Reply:
I added a note that `is_bst` and `is_balanced_tree` functions are available to them. I don't believe the tests they write here are actually ever run (verified by going into Learn and writing junk code in here). Since the Test Suite only runs the `test_always_pass` test, anything they write in here should be able to run.

That being said, I would love for them to think about how these tests are written, so I also added the `is_bst`, `height`, and `is_balanced_tree` functions to the example tests. I'm hopeful that'll make it super clear what is happening in those example tests.

Lemme know if anything seems amiss!